### ISSUE TITLE
Why is the function 'torch.topk' inconsistent on cpu and cuda?

### ISSUE THREAD

**Issue by Imsovegetable** [pytorch/pytorch#70234]
### 🐛 Describe the bug

The function 'torch.topk' will return different results when the input tensor is on cpu and cuda.
I am trying to calculate the top-k accuracy for each row in a matrix. So I typed in like this:
```python
import torch

b = torch.rand([4,3])
c = torch.rand([3,4])
a = torch.mm(b, c)
_a = torch.mm(b, c).cuda()

torch.topk(a, k=a.size(0), sorted=False)[1]
torch.topk(_a, k=_a.size(0), sorted=False)[1]
```
and then the same function will return different result as below
![image](https://user-images.githubusercontent.com/61308791/146899843-9c5aa02c-eea2-4e6a-8697-ae6b2fa554f5.png)

Obviously the result on cuda is completely wrong! 
Please fix it so it will not puzzle anyone else and waste their time.

### Versions

Collecting environment information...
PyTorch version: 1.10.0+cu113
Is debug build: False
CUDA used to build PyTorch: 11.3
ROCM used to build PyTorch: N/A

OS: Microsoft Windows 10 家庭中文版
GCC version: Could not collect
Clang version: Could not collect
CMake version: Could not collect
Libc version: N/A

Python version: 3.9.5 (default, May 18 2021, 14:42:02) [MSC v.1916 64 bit (AMD64)] (64-bit runtime)
Python platform: Windows-10-10.0.19042-SP0
Is CUDA available: True
CUDA runtime version: 11.3.109
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3060 Laptop GPU
Nvidia driver version: 496.49
cuDNN version: C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.3\bin\cudnn_ops_train64_8.dll
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] mypy==0.910
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.21.2
[pip3] torch==1.10.0+cu113
[pip3] torchmetrics==0.6.0
[pip3] torchvision==0.11.1+cu113
[conda] blas                      1.0                         mkl    http://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/free
[conda] mkl                       2021.4.0           haa95532_640    http://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/main
[conda] mkl-service               2.4.0            py39h2bbff1b_0    http://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/main
[conda] mkl_fft                   1.3.1            py39h277e83a_0    http://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/main
[conda] mkl_random                1.2.2            py39hf11a4ad_0    http://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/main
[conda] mypy                      0.910                    pypi_0    pypi
[conda] mypy-extensions           0.4.3                    pypi_0    pypi
[conda] numpy                     1.21.2           py39hfca59bb_0    http://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/main
[conda] numpy-base                1.21.2           py39h0829f74_0    http://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/main
[conda] torch                     1.10.0+cu113             pypi_0    pypi
[conda] torchmetrics              0.6.0                    pypi_0    pypi
[conda] torchvision               0.11.1+cu113             pypi_0    pypi

cc @ezyang @gchanan @zou3519 @bdhirsh

**Comment by ngimel** [comment]
Why do you think the values produced by cuda are wrong? `sorted=False` argument does not return the elements in original order, nor in any particular order. The docs say `"The boolean option sorted if True, will make sure that the returned k elements are themselves sorted"`, if `sorted` is False, top k elements can be returned in any order. Both cuda and cpu implementations are doing that.

**Comment by gchanan** [comment]
Here is a good description: https://github.com/pytorch/pytorch/issues/60300

Some docs suggestions I think would clarify things:

The current docs say:
> The boolean option sorted if True, will make sure that the returned k elements are themselves sorted`

It's a little unclear whether this refers to the indices and what (if any) guarantees are present if the sorted = False.  Assuming that we make no guarantees as to the sorted-ness of the indices (is that true?!) something like:

> If `sorted` is `True`, the `values` will be returned in sorted order according to `largest`.  Otherwise, there are no guarantees as to the order of the returned `values`.  There are no guarantees as to the order of the returned `indices` whether or not `sorted` is True or False.

**Comment by ngimel** [comment]
Indices correspond to the values, e.g. if we return `((a,b,c), (i0,i1,i2))` it means that value `a` was at position `i0`, `b` at position `i1` and `c` at position `i2` in the original tensor. Thus, we don't make any independent guarantees on the sorted-ness of the indices, indices order is defined by values order.

**Comment by gchanan** [comment]
Ok, updated:

> If `sorted` is `True`, the `values` will be returned in sorted order according to `largest`.  Otherwise, there are no guarantees as to the order of the returned `values`.  The `indices` order corresponds to the order of the `values` returned.

**Comment by Imsovegetable** [comment]
Sorry, I understood it. I used to mistakenly thought `sorted` means that the input tensor has already been sorted and that is my fault.
Thanks a lot!
